### PR TITLE
expression: add SubTime

### PR DESCRIPTION
**Pull request by hawkingrei**
Complete the exprssion `Subtime`
See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_subtime

### CODE REVIEW

**Comment by hawkingrei** [comment]
@shenli  @XuHuaiyu  PTAL

File: expression/builtin_time.go [hunk 1]
```diff
@@ -1813,6 +1813,49 @@ func strDurationAddDuration(d string, arg1 types.Duration) (result types.Datum,
...
+func strDurationSubDuration(d string, arg1 types.Duration) (result types.Datum, err error) {
+	arg0, err := types.ParseDuration(d, getFsp(d))
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	tmpDuration, err := arg0.Sub(arg1)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	if getFsp(d) != 0 {
```
Comment (shenli) [inline expression/builtin_time.go]: Could we get fsp from args0?

File: expression/builtin_time.go [hunk 2]
```diff
@@ -2372,7 +2415,63 @@ type builtinSubTimeSig struct {
...
+		arg1 = args[1].GetMysqlDuration()
+	default:
+		s, err := args[1].ToString()
+		if err != nil {
+			return d, errors.Trace(err)
+		}
+		if getFsp(s) == 0 {
+			arg1, err = types.ParseDuration(s, 0)
+		} else {
+			arg1, err = types.ParseDuration(s, types.MaxFsp)
```
Comment (shenli) [inline expression/builtin_time.go]: Why use MaxFsp instead of getFsp(s)?

Comment (hawkingrei) [inline expression/builtin_time.go]: @shenli 
if arg0 or arg1 is not  mysqltime or mysqlduration ,their fsp will be zero or the maxFsp.

mysql> select subtime(1,1.2);
+------------------+
| subtime(1,1.2)   |
+------------------+
| -00:00:00.200000 |
+------------------+
1 row in set (0.00 sec)


select subtime("1.2",1);
+------------------+
| subtime("1.2",1) |
+------------------+
| 00:00:00.200000  |
+------------------+
1 row in set (0.00 sec)

**Comment by shenli** [comment]
@XuHuaiyu PTAL

File: expression/typeinferer.go [hunk 3]
```diff
@@ -405,6 +405,17 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
 			tp = types.NewFieldType(mysql.TypeVarString)
 		}
 		tp.Charset, tp.Collate = types.DefaultCharsetForType(tp.Tp)
+	case ast.SubTime:
+		t := x.Args[0].GetType().Tp
+		switch t {
```
Comment (XuHuaiyu) [inline expression/typeinferer.go]: you can use 
``` go
switch t := x.Args[0].GetType().Tp; t{
}
```

File: expression/builtin_time.go [hunk 4]
```diff
@@ -1813,6 +1813,49 @@ func strDurationAddDuration(d string, arg1 types.Duration) (result types.Datum,
 	return
 }
 
+// strDatetimeSubDuration subs duration to datetime string, returns a datum value.
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: what does subs duration to datetime mean,
seems not correct.

File: expression/builtin_time.go [hunk 5]
```diff
@@ -1813,6 +1813,49 @@ func strDurationAddDuration(d string, arg1 types.Duration) (result types.Datum,
...
+	}
+	tmpDuration := arg0.Sub(&arg1time)
+	resultDuration, err := tmpDuration.ConvertToTime(mysql.TypeDatetime)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	if getFsp(d) != 0 {
+		tmpDuration.Fsp = types.MaxFsp
+	} else {
+		tmpDuration.Fsp = 0
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: s/0/types.MinFsp

**Comment by hawkingrei** [comment]
@XuHuaiyu  Fix

**Comment by XuHuaiyu** [comment]
LGTM
PTAL @tiancaiamao

File: expression/builtin_time.go [hunk 6]
```diff
@@ -2372,7 +2415,63 @@ type builtinSubTimeSig struct {
...
-	return d, errFunctionNotExists.GenByArgs("SUB_TIME")
+	args, err := b.evalArgs(row)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	if args[0].IsNull() || args[1].IsNull() {
+		return
+	}
+	var arg1 types.Duration
+	switch tp := args[1].Kind(); tp {
```
Comment (tiancaiamao) [inline expression/builtin_time.go]: `tp` is not used in switch body,  so you can just write
```
switch args[1].Kind() {
```

File: expression/builtin_time.go [hunk 7]
```diff
@@ -2372,7 +2415,63 @@ type builtinSubTimeSig struct {
...
+			arg1, err = types.ParseDuration(s, 0)
+		} else {
+			arg1, err = types.ParseDuration(s, types.MaxFsp)
+		}
+		if err != nil {
+			return d, errors.Trace(err)
+		}
+
+	}
+	switch tp := args[0].Kind(); tp {
```
Comment (tiancaiamao) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time_test.go [hunk 8]
```diff
@@ -568,6 +568,73 @@ func (s *testEvaluatorSuite) TestAddTimeSig(c *C) {
...
+		tmpInput := types.NewIntDatum(t.Input)
+		tmpInputDuration := types.NewIntDatum(t.InputDuration)
+		f, err := fc.getFunction(datumsToConstants([]types.Datum{tmpInput, tmpInputDuration}), s.ctx)
+		c.Assert(err, IsNil)
+		d, err := f.eval(nil)
+		c.Assert(err, IsNil)
+		result, _ := d.ToString()
+		c.Assert(result, Equals, t.expect)
+	}
+
```
Comment (tiancaiamao) [inline expression/builtin_time_test.go]: remove this empty line

File: expression/typeinferer.go [hunk 9]
```diff
@@ -395,8 +395,17 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
 		tp.Flag |= mysql.UnsignedFlag
 	// time related
 	case ast.AddTime:
-		t := x.Args[0].GetType().Tp
-		switch t {
+		switch t := x.Args[0].GetType().Tp; t {
```
Comment (tiancaiamao) [inline expression/typeinferer.go]: `t` is not used, so use
```
switch x.Args[0].GetType().Tp: {
```

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**